ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,553 @@
+// Copyright 2021-2023 Protocol Labs
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use crate::node::CollapsedNode;
+use crate::node::{Link, Node};
+use crate::Error;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::ser::Serialize;
+use fvm_ipld_encoding::CborStore;
+use serde::de::DeserializeOwned;
+use std::marker::PhantomData;
+
+impl<V, BS, Ver> crate::AmtImpl<V, BS, Ver>
+where
+    V: DeserializeOwned + Serialize,
+    BS: Blockstore,
+    Ver: crate::root::version::Version,
+{
+    /// Iterate over the AMT. Alternatively, you can directly iterate over the AMT without calling
+    /// this method:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// let kvs: Vec<u64> = (0..=5).collect();
+    /// kvs
+    ///     .iter()
+    ///     .map(|k| amt.set(u64::try_from(*k).unwrap(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// for kv in &amt {
+    ///     let (k, v) = kv?;
+    ///     println!("{k:?}: {v:?}");
+    /// }
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn iter(&self) -> Iter<'_, V, &BS, Ver> {
+        Iter {
+            stack: vec![IterStack {
+                node: Some(&self.root.node),
+                idx: 0,
+            }],
+            blockstore: &self.block_store,
+            bit_width: self.bit_width(),
+            ver: PhantomData,
+            key: 0,
+        }
+    }
+
+    /// Generates an iterator containing the values of the AMT, in sorted order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// amt.set(2, "two".to_owned());
+    /// amt.set(1, "one".to_owned());
+    ///
+    /// let vals = amt.values().collect::<Vec<_>>();
+    /// assert_eq!(vals, ["one", "two"]);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn values(&self) -> impl Iterator<Item = &V> {
```

Comment:
I'd drop both the keys/values methods. They're not all _that_ useful and they'll panic on error.